### PR TITLE
Fix bugs in several calendars with new continuity test

### LINKED ISSUE #2703: Fix negative numbers across icu_calendar
Poking around, I keep finding places in icu_calendar where we do the wrong thing with negative numbers. We need to fix this, since one value proposition of ICU4X's calendar crate is that we generate correct results across the range required by Temporal.

I made a branch, [iso-negative](https://github.com/unicode-org/icu4x/tree/iso-negative), with some test cases that are currently failing.

Actions on this ticket:

- [x] Clone my branch and change the code so that the tests I added can all pass
- [ ] Add tests to cover all operations that involve signed integers
- [ ] Fix code that these tests reveal to be broken, perhaps by using the new `div_rem_euclid` helper function (#2704)

Related: #2151

CC @pt2121 @Manishearth @pandusonu2

### LINKED ISSUE #2713: Fuzz the calendars crate
We have a lot of bugs like #2254, https://github.com/unicode-org/icu4x/pull/2704, https://github.com/unicode-org/icu4x/issues/2703

We should use [`cargo fuzz`](https://rust-fuzz.github.io/book/cargo-fuzz.html) to write fuzzers that:

 - Test that calendar conversions roundtrip
 - Test that calendar arithmetic APIs work

`cargo fuzz` is super easy to use and does expected type conversion for you; which ought to be highly useful for this.

For the roundtrip tests we can either write one test per calendar, or perhaps have an `Arbitrary`-deriving enum that selects between calendars (we could also use AnyCalendarKind for this but then it would have to impl `Arbitrary` behind a feature)

### LINKED ISSUE #4914: Panic when IslamicUmmAlQura month has 31 days
```rs
use icu::calendar::Calendar;
use icu::calendar::islamic::IslamicUmmAlQura;
use icu::calendar::types::{Era, MonthCode};

use std::str::FromStr;

fn main() {
  let cal = IslamicUmmAlQura::new_always_calculating();
  let era = Era::from_str("ah").unwrap();
  let year = -6823;
  let month_code = MonthCode::from_str("M01").unwrap();
  let dt = cal.date_from_codes(era, year, month_code, 1);
}
```

panics with
```
umm-al-qura months must not be more than 30 days
```

### PR DESCRIPTION
**Pull request by sffc**
Fixes #2703
Fixes #4914
Mostly fixes #2713

I wrote a test that checks for invariants on calendar behavior. In addition to ISO round-trip, it tests simple day-based calendar arithmetic, which is supported in all calendars, and it exercises a lot of code paths that aren't otherwise covered, such as month length, year length, and leap year status, which unveiled bugs in multiple calendars (Observational Islamic, Saudi Islamic, Hebrew, Chinese, Coptic, and Ethiopian).

There are several loosely related changes in this PR so I made an effort for it reviewable commit-by-commit.

### CODE REVIEW

File: utils/calendrical_calculations/src/chinese_based.rs [hunk 1]
```diff
@@ -269,7 +269,18 @@ pub(crate) fn midnight<C: ChineseBased>(moment: Moment) -> Moment {
 pub(crate) fn new_year_in_sui<C: ChineseBased>(prior_solstice: RataDie) -> (RataDie, RataDie) {
     // s1 is prior_solstice
     // Using 370 here since solstices are ~365 days apart
+    // Both solstices should fall in December
```
Comment (sffc) [inline utils/calendrical_calculations/src/chinese_based.rs:272]: This assertion fails for years before about 20,000 BCE. Is this WAI? I'm not sure which factor influences this more:

1. Although the Gregorian calendar drifts by less than Julian, it still drifts. Do we hit a threshold around 20,000 BCE where the solstice moved into November?
2. Or is this just an artifact of floating point error in astronomical calculations?
3. Or are we getting sidereal drift?

**Comment by sffc** [comment]
Tests are failing because I need to regen Chinese cached calendar data (does this break semver or is this new in 1.5?), but I want to wait until we discuss the above question.

**Comment by Manishearth** [comment]
I disagree with the commit changing the bounds of the Chinese calendar. Firstly, the calendar is not that old (even the thing about January 20 in 30 AD is dubious, but that at least works with the math).

More importantly, the way the Chinese new year is defined ("such that the winter solstice falls in the eleventh month") mathematically gives you a narrow range of ~30 days in which it may occur. I don't have the math at hand but when implementing this optimization I did the math and.

It also matches Aslaksen's _The Mathematics of the Chinese Calendar_.

This is a researched and mathematically accurate optimization and I do not think we should get rid of it because floating point error is breaking tests for date ranges that are irrelevant for the calendar.

**Comment by Manishearth** [comment]
It's no longer an optimization since we have unused space there but I'd still prefer to keep it around, if the computer is producing dates outside of that range the computer is probably wrong.

**Comment by sffc** [comment]
Please see my comment above regarding the Chinese calendar drift.

File: utils/calendrical_calculations/src/chinese_based.rs (hunk 1 above)
Comment (Manishearth) [reply utils/calendrical_calculations/src/chinese_based.rs:272]: This is WAI but we could make the assertion stop failing outside that range, or clamp the Chinese calendar.

AIUI the Gregorian calendar accounts for the precession of the equinoxes (the biggest factor here). There are some other smaller factors that lead to the 3216 number, but the thing is that at that scale the planetary bodies _are not that predictable_. Basically I'm not convinced there's a sensible way to actually make any predictions about dates that far off for something like this.

**Comment by Manishearth** [comment]
Ah, I was reviewing this on a phone and couldn't see all the comments. I am still leaning towards considering those as cases we shouldn't try to fix. I'm somewhat open to bumping up the new year offset range to allow for more dates, it doesn't cost us anything. We should still document this well and have assertions that between 31 AD and  maybe 2000 years from now, the new year falls in the narrower range specified, citing Reingold and Aslasken.

**Comment by sffc** [comment]
According to Wolfram Alpha ([query](https://www.wolframalpha.com/input?i=winter+solstice%2820000+BC%29+in+UTC)):

| Winter Solstice Time (UTC) | Winter Solstice Date |
|---|---|
| 2:41 pm LMT | Saturday, November 24, 20000 BC (extrapolated Gregorian calendar) |
| 11:27 am LMT | Saturday, November 28, 19000 BC (extrapolated Gregorian calendar) |
| 4:45 am LMT | Saturday, December 1, 18000 BC (extrapolated Gregorian calendar) |
| 4:27 pm LMT | Friday, December 4, 17000 BC (extrapolated Gregorian calendar) |
| 8:57 pm LMT | Thursday, December 6, 16000 BC (extrapolated Gregorian calendar) |
| 5:09 pm LMT | Wednesday, December 9, 15000 BC (extrapolated Gregorian calendar) |
| 4:16 am LMT | Tuesday, December 11, 14000 BC (extrapolated Gregorian calendar) |
| 6:30 am LMT | Sunday, December 13, 13000 BC (extrapolated Gregorian calendar) |
| 12:16 am LMT | Friday, December 14, 12000 BC (extrapolated Gregorian calendar) |
| 10:50 am LMT | Tuesday, December 15, 11000 BC (extrapolated Gregorian calendar) |
| 3:31 pm LMT | Saturday, December 15, 10000 BC (extrapolated Gregorian calendar) |
| 4:09 pm LMT | Wednesday, December 16, 9000 BC (extrapolated Gregorian calendar) |
| 2:22 pm LMT | Sunday, December 16, 8000 BC (extrapolated Gregorian calendar) |
| 11:41 am LMT | Thursday, December 17, 7000 BC (extrapolated Gregorian calendar) |
| 9:18 am LMT | Monday, December 17, 6000 BC (extrapolated Gregorian calendar) |
| 7:58 am LMT | Friday, December 18, 5000 BC (extrapolated Gregorian calendar) |
| 8:15 am LMT | Tuesday, December 18, 4000 BC (extrapolated Gregorian calendar) |
| 9:47 am LMT | Saturday, December 19, 3000 BC (extrapolated Gregorian calendar) |
| 12:16 pm LMT | Wednesday, December 19, 2000 BC (extrapolated Gregorian calendar) |
| 2:50 pm LMT | Sunday, December 20, 1000 BC (extrapolated Gregorian calendar) |
| 4:21 pm LMT | Thursday, December 20, 1 AD (extrapolated Gregorian calendar) |
| 9:54 am LMT | Sunday, December 21, 1000 AD (extrapolated Gregorian calendar) |
| 5:21 am PST | Thursday, December 21, 2000 |
| 8:18 pm PST | Sunday, December 21, 3000 |
| 5:21 am PST | Thursday, December 21, 4000 |
| 7:44 am PST | Sunday, December 21, 5000 |
| 3:02 am PST | Wednesday, December 20, 6000 |
| 2:59 pm PST | Friday, December 19, 7000 |
| 7:57 pm PST | Sunday, December 17, 8000 |
| 6:36 pm PST | Tuesday, December 16, 9000 |
| 12:01 pm PST | Thursday, December 14, 10000 |
| 1:28 am PST | Saturday, December 13, 11000 |
| 12:14 pm PST | Sunday, December 10, 12000 |
| 9:42 pm PST | Monday, December 8, 13000 |
| 6:41 am PST | Wednesday, December 6, 14000 |
| 3:53 pm PST | Thursday, December 4, 15000 |
| 1:10 am PST | Saturday, December 2, 16000 |
| 10:11 am PST | Sunday, November 30, 17000 |
| 5:55 pm PST | Monday, November 27, 18000 |
| 11:13 pm PST | Tuesday, November 25, 19000 |
| 12:47 am PST | Thursday, November 23, 20000 |

**Comment by sffc** [comment]
The problem is that everything currently ends up calculating a `PackedChineseBasedYearInfo` which has this requirement about the allowed start dates in Jaunary/February, but it's trivial to break that requirement, as shown above.

Should we use a different non-packed intermediate type without so many constraints?

**Comment by sffc** [comment]
The range of dates Temporal supports is from -100 million to 100 million days on either side of January 1 1970 (about 273000 AD).

**Comment by Manishearth** [comment]
> According to Wolfram Alpha ([query](https://www.wolframalpha.com/input?i=winter+solstice%2820000+BC%29+in+UTC)):

I understand this.  I don't think that affects what I said: These dates are beyond precise predictability and it is not meaningful to talk about them in this way. There is _probably_ going to be a shift in the solstice of that form, but I don't consider those dates as important for the Chinese calendar in the first place.

> Should we use a different non-packed intermediate type without so many constraints?

No, we should not introduce a separate code path. I am somewhat open to relaxing the constraints provided it doesn't have a perf impact (and has the strong assertions for the _actual_ date ranges we care about), but we should not introduce multiple codepaths in calendar code (once a `Date` has been constructed) since that makes it so much easier for things to subtly break.

**Comment by sffc** [comment]
> > According to Wolfram Alpha ([query](https://www.wolframalpha.com/input?i=winter+solstice%2820000+BC%29+in+UTC)):
> 
> I understand this. I don't think that affects what I said: These dates are beyond precise predictability and it is not meaningful to talk about them in this way. There is _probably_ going to be a shift in the solstice of that form, but I don't consider those dates as important for the Chinese calendar in the first place.

I do care about making things work for all dates in the range supported by Temporal. They don't have to be perfectly accurate, but the data model should support them.

Even if I increase the date offset to 64 days, we only get to about 20,000 AD/BC which is still only about 10% of the range required by Temporal.

> > Should we use a different non-packed intermediate type without so many constraints?
> 
> No, we should not introduce a separate code path. I am somewhat open to relaxing the constraints provided it doesn't have a perf impact (and has the strong assertions for the _actual_ date ranges we care about), but we should not introduce multiple codepaths in calendar code (once a `Date` has been constructed) since that makes it so much easier for things to subtly break.

I don't want to introduce another code path, but currently we use `PackedChineseBasedYearInfo` in the `DateInner`'s `ChineseBasedYearInfo`. I was proposing that we could instead use the packed year info in the data provider but then expand its fields into `ChineseBasedYearInfo`, increasing its stack size a little bit but removing the constraint on the new years offset, and for calculated data we go directly into the unpacked struct.

**Comment by sffc** [comment]
Brainstorming a few options:

1. Keep the invariants as-is, debug-panic when violated, and fall back to January 19 as the new year, which may or may not be a new moon. (current behavior)
2. Keep the invariants as-is, debug-panic when violated, and hack the Winter Solstice to always land on December 20/21, so that the new year contributes to be on a new moon.
3. Use an unpacked date inner type without the constraints, and use whatever garbage we get from the equinox function to calculate the new year. Note that it's possible to end up with a negative offset from January 1 when the solstice starts drifting into November or earlier.
4. Fall back to Proleptic Gregorian dates and eras when outside of a max supported range that we define for the Chinese calendar, such as 1000 BC to 5000 AD (±3000 years). -- this could be applied to other calendars as well.

**Comment by Manishearth** [comment]
Right, I understand your proposal, I don't want to increase the stack size of the Date type here.

I think the status quo with an explicit clamp for the solstice and no panic when the years are out of "bounds" is my ideal choice, and I don't think it breaks the data model. However I think expanding to Jan 1 is acceptable too.

In general I am fine with restricting our assertions to year ranges we consider relevant.

File: utils/calendrical_calculations/src/chinese_based.rs [hunk 2]
```diff
@@ -264,42 +272,78 @@ pub(crate) fn midnight<C: ChineseBased>(moment: Moment) -> Moment {
...
-    // Both solstices should fall in December
+    // Both solstices should fall in December, and the first solstice MUST fall on
+    // December 20, 21, or 22. It doesn't seem to change the outcome if the following
+    // solstice is out of range.
+    let (prior_solstice, drift_delta) = {
+        let adjusted = bind_winter_solstice::<C>(prior_solstice);
+        (adjusted, DriftDelta(adjusted - prior_solstice))
+    };
     let following_solstice = winter_solstice_on_or_before::<C>(prior_solstice + 370); // s2
-    debug_assert_eq!(
```
Comment (Manishearth) [inline utils/calendrical_calculations/src/chinese_based.rs]: issue: keep these asserts for modern year ranges (±2000 years?)

**Review by Manishearth — CHANGES_REQUESTED** [review]
I think I prefer the previous solution to this: I was expecting we could bind things without tweaking the algorithm too much. Let's stick to that, but not remove any asserts (just tweak them to only fire for modern year ranges).

File: utils/calendrical_calculations/src/chinese_based.rs (hunk 2 above)
Comment (sffc) [reply utils/calendrical_calculations/src/chinese_based.rs]: The asserts are obsolete if I use the `bind` functions.

**Comment by sffc** [comment]
Okay, as demonstrated in #4929, increasing the size of the offset above 32 bits is unavoidable.

Given the increased bits, I've now implemented a solution that pins the winter solstice to December 20-23 and allows the new year to land as early as January 19, which is the earliest the second New Moon could occur after December 20. It's not actually possible for a real-life Chinese New Year to land on January 19, because if the Winter Solstice were December 20 and there was a 29-day month starting on December 21, either it or the month after it would be a leap month. However, since we pin the winter solstice, it is possible that there are no winter leap months, and therefore January 19 ends up being a possible output of the algorithm.

This solution is still better than the original solution in this PR because it works for a longer range of years and should theoretically be able to work for an arbitrarily long range.

**Comment by sffc** [comment]
I already pulled out some of the changes, but if desired, I can split this PR up further.

File: utils/calendrical_calculations/src/islamic.rs [hunk 3]
```diff
@@ -39,33 +41,61 @@ pub trait IslamicBasedMarker {
...
-        let new_rd = Self::fixed_from_islamic(extended_year + 1, 1, 1);
-        let diff = new_rd - prev_rd;
-        debug_assert!(
-            diff == 29 || diff == 30,
-            "Found extended year {extended_year} with month length {diff} for month 12"
-        );
-        if new_rd - prev_rd == 30 {
-            lengths[11] = true;
+        });
+        // To maintain invariants for calendar arithmetic, if astronomy finds
```
Comment (Manishearth) [inline utils/calendrical_calculations/src/islamic.rs:88]: praise: thanks, i was considering a technique like this

**Review by Manishearth — CHANGES_REQUESTED** [review]
Regarding the chinese fix: I'm okay both with the current solution and the previous one of using more unused PackedYearInfo bits (provided no stack or data size increases) to set FIRST_NY to 1 and not do any clamping. I'm not sure which I prefer, I think I actually slightly prefer the original approach given that Gregorian drift is real, but either is fine.

File: components/calendar/src/islamic.rs [hunk 4]
```diff
@@ -221,31 +221,34 @@ impl IslamicTabular {
 #[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
 pub(crate) struct IslamicYearInfo {
     packed_data: PackedIslamicYearInfo,
-    /// Is the previous year 355 days (short = 354)
-    prev_year_long: bool,
+    prev_year_length: u16,
```
[the remark this answers is not in the document]
Comment (sffc) [reply components/calendar/src/islamic.rs:224]: It fails for various years in the islamic-observational calendar in the range under test. I think it always too short (353 days). If you want to keep the bits small, I can compress this down to a 3 or 4 value enum.

[the remark this answers is not in the document]
Comment (Manishearth) [reply components/calendar/src/islamic.rs:224]: that would be fine. However, can you post the actual years it fails for?

[the remark this answers is not in the document]
Comment (Manishearth) [reply components/calendar/src/islamic.rs:224]: and ideally, I would like to have that assertion stick around for the other calendars

[the remark this answers is not in the document]
Comment (sffc) [reply components/calendar/src/islamic.rs:224]:
```
2024-05-22T23:12:39.835Z WARN  [calendrical_calculations::islamic] (ObservationalIslamic) Found year -269 AH with length 353
2024-05-22T23:12:40.852Z WARN  [calendrical_calculations::islamic] (ObservationalIslamic) Found year -40 AH with length 353
2024-05-22T23:12:41.775Z WARN  [calendrical_calculations::islamic] (ObservationalIslamic) Found year 168 AH with length 353
2024-05-22T23:12:43.965Z WARN  [calendrical_calculations::islamic] (ObservationalIslamic) Found year 670 AH with length 353
2024-05-22T23:12:44.374Z WARN  [calendrical_calculations::islamic] (ObservationalIslamic) Found year 763 AH with length 353
2024-05-22T23:12:45.868Z WARN  [calendrical_calculations::islamic] (ObservationalIslamic) Found year 1107 AH with length 353
2024-05-22T23:12:49.495Z WARN  [calendrical_calculations::islamic] (SaudiIslamic) Found year 454 AH with length 353
2024-05-22T23:12:52.285Z WARN  [calendrical_calculations::islamic] (SaudiIslamic) Found year 669 AH with length 353
2024-05-22T23:12:55.263Z WARN  [calendrical_calculations::islamic] (SaudiIslamic) Found year 899 AH with length 353
```

[the remark this answers is not in the document]
Comment (sffc) [reply components/calendar/src/islamic.rs]: I pushed code that retains the assertion in `calendrical_calculations`, and `icu_calendar` is fine with all three year lengths.

[the remark this answers is not in the document]
Comment (Manishearth) [reply components/calendar/src/islamic.rs]: Hmm, thanks. these years are in practical range for these calendars so we should ideally figure out what's wrong with the math but til then it's fine to hack it this way.

File: components/calendar/src/islamic.rs [hunk 5]
```diff
@@ -218,10 +218,41 @@ impl IslamicTabular {
     }
 }
 
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+enum IslamicYearLength {
+    L355,
+    L354,
+    L353,
```
Comment (Manishearth) [inline components/calendar/src/islamic.rs:225]: nit: document this as an edge case that shouldn't actually occur (and list the positive years/cals for which it does occur)

File: utils/calendrical_calculations/src/chinese_based.rs [hunk 6]
```diff
@@ -278,7 +291,39 @@ pub(crate) fn new_year_in_sui<C: ChineseBased>(prior_solstice: RataDie) -> (Rata
...
+    let (iso_year, iso_month, iso_day) = match iso_from_fixed(solstice) {
+        Ok(ymd) => ymd,
+        Err(_) => {
+            debug_assert!(false, "Solstice REALLY out of bounds: {solstice:?}");
+            return solstice;
+        }
+    };
+    if iso_month < 12 || iso_day < 20 {
+        #[cfg(feature = "logging")]
+        log::trace!("({}) Solstice out of bounds: {solstice:?}", C::DEBUG_NAME);
```
Comment (Manishearth) [inline utils/calendrical_calculations/src/chinese_based.rs:308]: nit: we should upgrade this to a debug assertion for iso_years that are ~±2000years.

File: components/calendar/src/islamic.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sffc) [reply components/calendar/src/islamic.rs]: Opened https://github.com/unicode-org/icu4x/issues/4930 and I'll add references in the code

File: utils/calendrical_calculations/src/chinese_based.rs (hunk 6 above)
Comment (sffc) [reply utils/calendrical_calculations/src/chinese_based.rs]: Done

File: components/calendar/src/islamic.rs (hunk 5 above)
Comment (sffc) [reply components/calendar/src/islamic.rs:232]: Done

**Review by Manishearth — APPROVED** [review]
I kind of feel that the original Jan 1 changes were better, at this point (sorry, I've gotten turned around by the original arguments and the data) but let's first land this.

**Comment by sffc** [comment]
Branch with individual commits archived at https://github.com/sffc/omnicu/tree/archive/negative-tests

**Pull request merged**